### PR TITLE
User can register and login

### PR DESCRIPTION
**Pull request by YKo20010**

### CODE REVIEW

File: acai-ios/TableViewCells/InputTableViewCell.swift [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    var invalidEntryLabel: UILabel!
+    var textField: UITextField!
+
+    // MARK: Constraint Constants
+    private enum FileConstants {
+        static let lineHeight = 1
+        static let leadingOffset = 35
+        static let trailingOffset = 22
+        static let verticalOffset = 4
+
```
Comment (kevinchan159) [inline acai-ios/TableViewCells/InputTableViewCell.swift:25]: delete line

File: acai-ios/ViewControllers/LoginViewController.swift [hunk 2]
```diff
@@ -122,71 +108,184 @@ class LoginViewController: UIViewController {
...
 
-    func switchDataEntry(to type: EntryType) {
-        if type == .login {
-            emailInputView.snp.remakeConstraints { make in
-                make.leading.equalToSuperview().offset(FileConstants.leadingOffset)
-                make.trailing.equalToSuperview().offset(-FileConstants.trailingOffset)
-                make.height.equalTo(FileConstants.inputViewHeight)
-                make.top.equalTo(view.safeAreaLayoutGuide.snp.top).offset(FileConstants.inputViewTopOffset)
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        return inputItems.count + 1
```
Comment (kevinchan159) [inline acai-ios/ViewControllers/LoginViewController.swift:98]: maybe add a comment about why you do `+ 1`

File: acai-ios/ViewControllers/LoginViewController.swift [hunk 3]
```diff
@@ -122,71 +108,184 @@ class LoginViewController: UIViewController {
...
+    }
+
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell = indexPath.row == inputItems.count ? tableView.dequeueReusableCell(withIdentifier: FileConstants.roundedButtonCellReuseIdentifier, for: indexPath) as! RoundedButtonTableViewCell : tableView.dequeueReusableCell(withIdentifier: FileConstants.inputCellReuseIdentifier, for: indexPath) as! InputTableViewCell
+        if let cell = cell as? InputTableViewCell {
```
Comment (kevinchan159) [inline acai-ios/ViewControllers/LoginViewController.swift]: instead of doing this check (`cell = cell as? InputTableViewCell`) I think its better to just do
```
if indexPath.row == inputItems.count {
   let cell = ...
    ...
    return cell
}
let cell = ...
...
return cell
```

File: acai-ios/ViewControllers/LoginViewController.swift [hunk 4]
```diff
@@ -122,71 +108,184 @@ class LoginViewController: UIViewController {
...
+}
+
+extension LoginViewController: SelectRoundedButtonDelegate {
+    func register() {
+        attemptedNetworking = true
+        (tableView.cellForRow(at: IndexPath(row: 0, section: 0)) as? InputTableViewCell)?.invalidEntryLabel.isHidden = firstName.isValidName()
+        (tableView.cellForRow(at: IndexPath(row: 1, section: 0)) as? InputTableViewCell)?.invalidEntryLabel.isHidden = lastName.isValidName()
+        (tableView.cellForRow(at: IndexPath(row: 2, section: 0)) as? InputTableViewCell)?.invalidEntryLabel.isHidden = phoneNumber.isValidPhoneNumber()
+        (tableView.cellForRow(at: IndexPath(row: 3, section: 0)) as? InputTableViewCell)?.invalidEntryLabel.isHidden = email.isValidEmail()
+        (tableView.cellForRow(at: IndexPath(row: 4, section: 0)) as? InputTableViewCell)?.invalidEntryLabel.isHidden = password.isValidPassword()
```
Comment (kevinchan159) [inline acai-ios/ViewControllers/LoginViewController.swift]: lines 212-215 is sort of hard to read. can we have unwrap each one
```
if let cell = tableView.cellForRow(at: IndexPath(row: 0, section: 0)) as? InputTableViewCell {
    ...
}
...
```
Also i don't think this handles the case where the cell isn't on the screen. You might want to double check that this doesn't cause any issues

File: acai-ios/ViewControllers/LoginViewController.swift [hunk 5]
```diff
@@ -122,71 +108,184 @@ class LoginViewController: UIViewController {
...
-                getUser(email: email, password: password).observe { (result) in
-                    switch result {
-                    case .value(let response):
-                        print(response)
-                    case .error(let error):
-                        print(error)
-                    }
+    func login() {
+        attemptedNetworking = true
+        guard let email = (tableView.cellForRow(at: IndexPath(row: 0, section: 0)) as? InputTableViewCell)?.textField.text, let password = (tableView.cellForRow(at: IndexPath(row: 1, section: 0)) as? InputTableViewCell)?.textField.text else { return }
```
Comment (kevinchan159) [inline acai-ios/ViewControllers/LoginViewController.swift]: split this into two lines:
```
guard let email = ...,
  let password = ... else { 
    return 
}
```

File: acai-ios/ViewControllers/LoginViewController.swift [hunk 6]
```diff
@@ -122,71 +108,184 @@ class LoginViewController: UIViewController {
...
+        (tableView.cellForRow(at: IndexPath(row: 0, section: 0)) as? InputTableViewCell)?.invalidEntryLabel.isHidden = email.isValidEmail()
+        (tableView.cellForRow(at: IndexPath(row: 1, section: 0)) as? InputTableViewCell)?.invalidEntryLabel.isHidden = password != ""
```
Comment (kevinchan159) [inline acai-ios/ViewControllers/LoginViewController.swift]: same comment about unwrapping each one since its hard to read

**Review by kevinchan159 — CHANGES_REQUESTED** [review]
amazing work!!!

File: acai-ios/ViewControllers/LoginViewController.swift (hunk 4 above)
Comment (YKo20010) [reply acai-ios/ViewControllers/LoginViewController.swift]: The array of input cells is hard coded and register() can only be called from the register screen where there are 5 input cells

**Review by kevinchan159 — DISMISSED** [review]
looks awesome

File: Podfile.lock [hunk 7]
```diff
@@ -8,10 +8,10 @@ PODS:
   - IGListKit/Default (3.4.0):
     - IGListKit/Diffing
   - IGListKit/Diffing (3.4.0)
-  - Kingfisher (5.1.1)
+  - Kingfisher (5.3.1)
```
Comment (doctor3w) [inline Podfile.lock]: please change these back unless the new versions are necessary. Pod updates should be a separate commit

File: acai-ios.xcodeproj/xcshareddata/xcschemes/acai-ios.xcscheme [hunk 8]
```diff
@@ -42,7 +42,7 @@
       </AdditionalOptions>
    </TestAction>
    <LaunchAction
-      buildConfiguration = "Release"
+      buildConfiguration = "Debug"
```
Comment (doctor3w) [inline acai-ios.xcodeproj/xcshareddata/xcschemes/acai-ios.xcscheme]: Please change this release scheme back to release. We have a debug scheme if you need to use the dev URLs

File: acai-ios/AppDelegate.swift [hunk 9]
```diff
@@ -36,7 +36,7 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
 
     func setupWindow() {
         window = UIWindow()
-        window?.rootViewController = UINavigationController(rootViewController: MenuViewController())
```
Comment (doctor3w) [inline acai-ios/AppDelegate.swift:39]: Change this back to Menu

File: acai-ios/Endpoints.swift [hunk 10]
```diff
@@ -11,15 +11,29 @@ import FutureNova
...
+    }
+
+    private struct LoginBody: Codable {
+        let email: String
+        let password: String
+    }
+
     /// The login endpoint to get a User
     static func login(email: String, password: String) -> Endpoint {
+        let body = LoginBody(email: email, password: password)
```
Comment (doctor3w) [inline acai-ios/Endpoints.swift:29]: looks good!

File: acai-ios/TableViewCells/InputTableViewCell.swift [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+        textField.attributedPlaceholder = NSAttributedString(string: placeholder, attributes: attr)
+        label.text = type.rawValue
+        invalidEntryLabel.text = "Invalid \(type.rawValue.lowercased()) entered"
+        switch type {
+        case .email:
+            textField.autocapitalizationType = .none
+            textField.isSecureTextEntry = false
+            textField.keyboardType = .emailAddress
+        case .firstName:
+            textField.autocapitalizationType = .words
```
Comment (doctor3w) [inline acai-ios/TableViewCells/InputTableViewCell.swift:78]: Can we turn off autocorrect for names as well?

File: acai-ios/TableViewCells/InputTableViewCell.swift [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+        case .phoneNumber:
+            textField.autocapitalizationType = .none
+            textField.isSecureTextEntry = false
+            textField.keyboardType = .phonePad
+        }
+    }
+
+    private func setupConstraints() {
+        invalidEntryLabel.snp.makeConstraints { make in
+            make.leading.equalToSuperview().offset(FileConstants.leadingOffset)
```
Comment (doctor3w) [inline acai-ios/TableViewCells/InputTableViewCell.swift]: If these constants are only used here we can just put them in the setupConstraints function

File: acai-ios/ViewControllers/LoginViewController.swift [hunk 13]
```diff
@@ -122,71 +108,191 @@ class LoginViewController: UIViewController {
...
-            nameInputView.invalidEntryLabel.isHidden = name.isValidName()
-            passwordInputView.invalidEntryLabel.isHidden = password.isValidPassword()
-            if email.isValidEmail() && name.isValidName() && password.isValidPassword() {
-                createUser(name: name, email: email, password: password)
+            return FileConstants.inputViewHeight
+        }
+    }
+
+    func setupKeyboard() {
+        NotificationCenter.default.addObserver(self, selector: #selector(showKeyboard), name: UIResponder.keyboardWillShowNotification, object: nil)
```
Comment (doctor3w) [inline acai-ios/ViewControllers/LoginViewController.swift]: Can't we achieve this effect with UITextFieldDelegate? NotificationCenter seems overkill for this

Comment (doctor3w) [reply acai-ios/ViewControllers/LoginViewController.swift]: Also I think if we make this a UITableViewController this keyboard logic will be done for us

File: acai-ios/Views/InputItem.swift [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+            return "Cathy"
+        case .password:
+            return "••••••••••••••"
+        case .phoneNumber:
+            return "8005554444"
+        }
+    }
+}
+
+class InputItem {
```
Comment (doctor3w) [inline acai-ios/Views/InputItem.swift]: Can we use a struct for this?

Comment (doctor3w) [inline acai-ios/Views/InputItem.swift]: InputItem is also a pretty generic name, could we make it more specific or namespace it with the LoginViewController?

File: Pods/Kingfisher/Sources/Cache/DiskStorage.swift [hunk 15]
```diff
@@ -196,11 +196,18 @@ public enum DiskStorage {
         }
 
         func cacheFileName(forKey key: String) -> String {
```
Comment (doctor3w) [inline Pods/Kingfisher/Sources/Cache/DiskStorage.swift]: Make sure to rerun pod install once you change back to the old versions (except for FutureNova) so these changes don't appear anymore.

**Review by doctor3w — CHANGES_REQUESTED** [review]

File: acai-ios/TableViewCells/InputTableViewCell.swift (hunk 11 above)
Comment (YKo20010) [reply acai-ios/TableViewCells/InputTableViewCell.swift:78]: I have autocorrect off for all textfields in the init(). It's autoCapitalization that is in the case statements.

File: acai-ios/Extensions/String+Extension.swift [hunk 16]
```diff
@@ -2,52 +2,30 @@
...
     }
 
-    // TODO: update name requirements
     func isValidName() -> Bool {
-        return self.count > 0
+        let nameRE = "[a-zA-Z]+"
+        return NSPredicate(format: "SELF MATCHES %@", nameRE).evaluate(with: self)
     }
 
     // TODO: update password requirements
```
Comment (cornell-appdev) [inline acai-ios/Extensions/String+Extension.swift:22]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/cuappdev/acai-ios/blob/77d46e0f2f8ea3ebec486458ed560955ce23fec3/acai-ios/Extensions/String+Extension.swift#L22"></span>TODOs should be resolved (update password requirements).<br />
<code>todo</code> <code>String+Extension.swift:22</code></td>
    </tr>
  </tbody>
</table>

**Comment by cornell-appdev** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Big PR, try to keep changes smaller if you can</td>
    </tr>
  </tbody>
</table>

### SwiftLint found issues

#### Warnings

File | Line | Reason |
| --- | ----- | ----- |
String+Extension.swift | 22 | TODOs should be resolved (update password requirements). (todo)


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>
